### PR TITLE
loadPlops() method

### PR DESCRIPTION
**Pull request by tkambler**
This PR introduces a backwards-compatible feature - the `loadPlops()` method. The concept is similar to that of Grunt's [loadTasks()](http://gruntjs.com/api/grunt.task#grunt.task.loadtasks) method. This feature allows you to decompose your plop structure into a series of multiple files, instead of one monolithic config file.

Example:

```
// plopfile.js
module.exports = function(plop) {
    plop.loadPlops('./plops');
};

// ./plops/my-generator.js
module.exports = function(plop) {
    plop.setGenerator('my-generator', {
        // ...
    });
};
```

### CODE REVIEW

**Comment by amwmedia** [comment]
Hi Tim, thanks for getting involved with plop! 

Because plopfiles are just node modules, isn't this basically possible today by doing something like the following?

``` javascript
module.exports = function (plop) {
  require('./plops/my-generator')(plop);
}
```

Using something like `loadPlops` is an interesting idea, and I see how it may make it easier to add new generators in the future as you develop. But I don't think it gets you any isolation that is not possible now through normal node development practices. Is that correct or am I missing something here?

**Comment by tkambler** [comment]
@amwmedia You're right - you can use the approach you describe. In fact, this is what one of my plop files currently looks like:

```
module.exports = function(plop) {

    const glob = require('glob');
    glob.sync('./plops/*.js').forEach((p) => {
        require(p)(plop);
    });

};
```

The advantage provided by `loadPlops()` is that you don't have to update your plopfile _every time_ you add or remove a module, and you don't have to write the boilerplate that you see above - it's just taken care of for you. The process of adding or removing a module is as simple as just adding or removing the module itself.

**Comment by tkambler** [comment]
@amwmedia You are correct, you could accomplish the same thing via `require()` as shown in your example. This is merely a nicety, but one that I've found to be very useful. I've found that it becomes more useful over time as you start adding more plops.

This is what one of my plopfiles currently looks like:

```
module.exports = function(plop) {

    const glob = require('glob');
    glob.sync('./plops/*.js').forEach((p) => {
        require(p)(plop);
    });

};
```

The `loadPlops()` method prevents you from having to write this boilerplate. Following the pattern shown in your example, it would look more like this:

```
module.exports = function (plop) {
    require('./plops/my-generator1')(plop);
    require('./plops/my-generator2')(plop);
    require('./plops/my-generator3')(plop);
    require('./plops/my-generator4')(plop);
    require('./plops/my-generator5')(plop);
    require('./plops/my-generator6')(plop);
    require('./plops/my-generator7')(plop);
    require('./plops/my-generator8')(plop);
    require('./plops/my-generator9')(plop);
    require('./plops/my-generator10')(plop);
}
```

**Comment by amwmedia** [comment]
I wonder if this would be better served as an npm module that I could reference in the readme? My concern is that by adding to the core plop code, it may give the impression that this is "the way" that plop generators are to be structured. If it's kept as an external library it is then a handy tool to help you structure your plop generators.

I'd probably use this myself as I think there's a lot of benefit to your approach. What do you think about that?

**Comment by tkambler** [comment]
If you want my two cents, it would be this (not trying to be snarky here): you'd be better off not doing it at all. I think the idea of creating a special "load-plop-tasks" module (see example below) exemplifies so much of what is wrong with npm right now. Modularization has been taken to a ridiculous extreme, and as a result, we've wound up with this massive pile of modules, many of which do some silly little thing that hardly justifies the module's existence in the first place.

```
// ./node_modules/load-plop-tasks/index.js - A new npm module for this - really?
module.exports = (plop, plopsPath) => {
    const glob = require('glob');
    glob.sync('*.js', {
        'cwd': plopsPath
    }).forEach((p) => {
        require(p)(plop);
    });
};

// ./plopfile.js
module.exports = function(plop) {
    require('load-plop-tasks')(plop, './plops');
};
```

Totally up to you, but IMO, the better approach would be to use the more basic `require('...')` approach in all of this library's core examples, thereby setting a simple, baseline approach by which all others can be judged. Then, in the documentation, simply make reference to the `loadPlops()` task. I believe that's what Grunt did, and it worked out fine for them IMO.

**Comment by amwmedia** [comment]
I'll give it some thought, thanks for your input and time 😄 . Please don't take my hesitancy as a slight on what you've done here, just trying to be cautious about how much is put into the base tool. I think one of plops biggest strengths is its simplicity. You don't lose simplicity overnight, it's lost one feature at a time. So I try to keep asking myself, does this belong as part of plop's core competency? Maybe the answer is "yes", I'm just not sure yet.

Sorry if that got a little overly cerebral haha.

**Comment by tkambler** [comment]
@amwmedia I totally know what you mean.

**Comment by amwmedia** [comment]
I've seen too many great, simple, tools that have become bloated with features that are only used by 10% of the users. But a newcomer has to weed through this huge API to find the pieces that they need.

File: mod/plop-base.js [hunk 1]
```diff
@@ -4,7 +4,10 @@ module.exports = (function () {
 	var inquirer = require('inquirer');
 	var handlebars = require('handlebars');
 	var changeCase = require('change-case');
+	var path = require('path');
+    var glob = require('glob');
```
Comment (laat) [inline mod/plop-base.js]: this file is indented with tabs, while you are using spaces.

**Comment by laat** [comment]
I think this is a great idea, I for one would like this feature :+1:

**Comment by tkambler** [comment]
Still thinking this one through, eh? ;)

**Comment by amwmedia** [comment]
leaning toward implementing it, just thinking it through a little more because long term the goal is to have npm installable "packs" of generators that you could pull into your plop file. I want to position this so that it could support that as well without changing the interface much or at all.

Also just been busy with work/life 😉

**Comment by amwmedia** [comment]
hang in there, I haven't forgotten about this. The release of node-plop has complicated this a bit though because this really needs to be implemented over there. I'm soon going to be pulling node-plop in as a dependency on this project so that the core logic only exists in one place.

**Comment by amwmedia** [comment]
I'll be shortly releasing a new version of plop that updates to the latest inquirer and also implements the core logic using node-plop. This will be a huge step forward for the project and will open up the door for a lot of features I've been wanting to implement. 

This PR represents one of those features, the ability to separate sets of generators into different files. Another part of that (as I see it) is being able to publish sets of generators as a node module and share them with others. This will be the first feature I add in after this next release, and I'm planning to complete it within the coming weeks (not months).

Thanks for your patients, I'm going to close this PR because the logic will need to live in [node-plop](https://github.com/amwmedia/node-plop). If you're interested with helping out on that project, please feel free to start an issue and we can discuss implementation ideas.

**Comment by amwmedia** [comment]
take a look at [plop.load](https://github.com/amwmedia/plop/blob/master/plop-load.md)

**Pull request closed without merging**